x/ibc/light-clients/07-tendermint/types/upgrade.go

```diff
@@ -26,102 +25,128 @@ import (
 //   and ProofSpecs do not match parameters set by committed client
 func (cs ClientState) VerifyUpgradeAndUpdateState(
 	ctx sdk.Context, cdc codec.BinaryMarshaler, clientStore sdk.KVStore,
-	upgradedClient exported.ClientState, upgradeHeight exported.Height, proofUpgrade []byte,
+	upgradedClient exported.ClientState, upgradedConsState exported.ConsensusState,
+	lastHeight exported.Height, proofUpgradeClient, proofUpgradeConsState []byte,
 ) (exported.ClientState, exported.ConsensusState, error) {
-	if cs.UpgradePath == "" {
+	if len(cs.UpgradePath) == 0 {
 		return nil, nil, sdkerrors.Wrap(clienttypes.ErrInvalidUpgradeClient, "cannot upgrade client, no upgrade path set")
 	}
-	upgradePath, err := constructUpgradeMerklePath(cs.UpgradePath, upgradeHeight)
-	if err != nil {
-		return nil, nil, sdkerrors.Wrapf(clienttypes.ErrInvalidUpgradeClient, "cannot upgrade client, unescaping key with URL format failed: %v", err)
-	}
 
-	// UpgradeHeight must be in same version as client state height
-	if cs.GetLatestHeight().GetVersionNumber() != upgradeHeight.GetVersionNumber() {
-		return nil, nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidHeight, "version at which upgrade occurs must be same as current client version. expected version %d, got %d",
-			cs.GetLatestHeight().GetVersionNumber(), upgradeHeight.GetVersionNumber())
+	if upgradedClient.GetLatestHeight().GetVersionNumber() <= cs.GetLatestHeight().GetVersionNumber() {
+		return nil, nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidHeight, "upgraded client height %s must be at greater version than current client height %s",
+			upgradedClient.GetLatestHeight(), cs.GetLatestHeight())
 	}
 
-	// UpgradeHeight must be greater than or equal to current client state height
-	if cs.GetLatestHeight().GetVersionHeight() > upgradeHeight.GetVersionHeight() {
-		return nil, nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidHeight, "version height at which upgrade occurs must be greater than or equal to current client height (%d > %d)",
-			cs.GetLatestHeight().GetVersionHeight(), upgradeHeight.GetVersionHeight(),
-		)
+	// UpgradeHeight must be the latest height of the client state as client must contain the consensus state
+	// of the last height, and should not contain a greater height before the upgrade.
+	if !cs.GetLatestHeight().EQ(lastHeight) {
```

Comment:
why do we need to pass in `lastHeight` if it equals the last height of the client state?

Reply:
good point!